src/quick-lint-js/parse.h

```diff
@@ -3392,6 +3409,26 @@ class parser {
     bool old_value_;
   };
 
+  class depth_guard {
+   public:
+    explicit depth_guard(parser *p, int old_depth_value) noexcept
+        : parser_(p), old_depth_value_(old_depth_value) {
+      p->depth_++;
+      if (p->depth_ > p->stack_limit) {
+        p->crash_on_unimplemented_token(__FILE__, __LINE__, __func__);
+      }
+    }
+
+    depth_guard(const depth_guard &) = delete;
+    depth_guard &operator=(const depth_guard &) = delete;
+
+    ~depth_guard() noexcept { this->parser_->depth_ = this->old_depth_value_; }
```

Comment:
Nit: Also assert that `this->parser_->depth_ == this->old_depth_value_+1;`, ensuring that depth changes are balanced?